tools/lkl/lib/virtio.c

```diff
@@ -67,16 +68,22 @@ void virtio_req_complete(struct virtio_req *req, uint32_t len)
 	struct virtio_queue *q = req->q;
 	struct virtio_dev *dev = req->dev;
 	uint16_t idx = le16toh(q->used->idx) & (q->num - 1);
-	int send_irq = 0;
+	uint16_t new = le16toh(q->used->idx) + 1;
 
 	q->used->ring[idx].id = htole16(req->idx);
 	q->used->ring[idx].len = htole16(len);
-	if (virtio_get_used_event(q) == q->used->idx)
-		send_irq = 1;
-	q->used->idx = htole16(le16toh(q->used->idx) + 1);
-
-	if (send_irq)
+	q->used->idx = htole16(new);
+
+	/* The condition is that only trigger the irq if
+	 * used->idx >= used_event + 1 > last_used_idx_signaled
+	 * i.e., the used_event set by virtio_net driver is between used->idx
+	 * and last_used_idx_signaled.
```

Comment:
I see. So it's saying "I just wrote to index `wrote_idx = q->used->idx` of the vring. The kernel wants me to tell him when I get up to index `used_event = virtio_get_used_event(q)`. So if `wrote_idx` is at least as big as the kernel wants to be notified about, *and* I haven't notified him since some time before `used_event`, I should let him know about this write, because it was at least as far as the point he was interested in."?

So I guess basically then the problem all along was that it was notifying the kernel about a write only when it wrote to *exactly* the point the kernel was interested in, rather than some point at least as far as what the kernel was interested in? And sometimes `q->used->idx` would be incremented twice in between calls to `virtio_req_complete` so it would miss the moment when it wanted to notify the kernel? This is great, I suspect this has been behind a lot of stability issues that popped up when making changes unrelated to virtio_net.

It might be worth going into a little more detail in the comment to explain the significance of `used->idx` and `last_used_idx_signaled` here --- I think the code around virtio_net is very confusing and it would help make it clearer to future maintainers. But LGTM, either way.